component/src/ibc/ibc_handler.rs

```diff
@@ -0,0 +1,189 @@
+/// IBC "app handlers" for the IBC component.
+///
+/// An app handler defines an interface for any IBC application to consume verified IBC events from
+/// the core IBC component. IBC applications listen to channel events that occur on the Port ID
+/// that they have subscribed to, and apply application-specific state transition logic.
+///
+/// The primary IBC application is the ICS20 transfer application, which allows for interchain
+/// token transfers.
+use anyhow::Result;
+use async_trait::async_trait;
+use ibc::core::ics04_channel::msgs::acknowledgement::MsgAcknowledgement;
+use ibc::core::ics04_channel::msgs::chan_close_confirm::MsgChannelCloseConfirm;
+use ibc::core::ics04_channel::msgs::chan_close_init::MsgChannelCloseInit;
+use ibc::core::ics04_channel::msgs::chan_open_ack::MsgChannelOpenAck;
+use ibc::core::ics04_channel::msgs::chan_open_confirm::MsgChannelOpenConfirm;
+use ibc::core::ics04_channel::msgs::chan_open_init::MsgChannelOpenInit;
+use ibc::core::ics04_channel::msgs::chan_open_try::MsgChannelOpenTry;
+use ibc::core::ics04_channel::msgs::recv_packet::MsgRecvPacket;
+use ibc::core::ics04_channel::msgs::timeout::MsgTimeout;
+use ibc::core::ics24_host::identifier::PortId;
+use std::collections::HashMap;
+
+/// AppHandlerCheck defines the interface for an IBC application to consume IBC channel and packet
+/// events, and apply their validation logic. This validation logic is used for stateful validation
+/// only.
+#[async_trait]
+pub trait AppHandlerCheck: Send + Sync {
+    async fn chan_open_init_check(&self, msg: &MsgChannelOpenInit) -> Result<()>;
```

Comment:
should this also have a `ctx` so that we can report back IBC events from IBC applications? I think probably yes